### PR TITLE
[PMAPI-1433] add campaign api resources

### PR DESCRIPTION
**Pull request by jharrell**
Fixes PMAPI-1433

## Checklist

- [x] Up to date with `main`
- [ ] All the tests are passing
  - [ ] Delete all resources created in tests
- [x] Prettier
- [x] Spectral Lint
- [x] `npm run bundle` outputs nothing suspect
- [x] `npm run postman` outputs nothing suspect

## Changes

- adds Campaign, Creative, and Upload routes to our docs.

## Areas of Concern (optional)

- testing, first time contributing.

### CODE REVIEW

File: lob-api-public.yml [hunk 1]
```diff
@@ -110,6 +110,12 @@ tags:
       Verify a list of US addresses.
       <div class="back-to-top" ><a href="#" onclick="toTopLink()">back to top</a></div>
 
+  - name: Campaigns
+    description: |
+      The campaigns endpoint allows you to create and view campaigns that can be used to send multiple letters or postcards.
+      The API provides endpoints for creating campaigns, updating campaigns, retrieving individual campaigns, and deleting
```
Comment (sudoku-lord) [inline lob-api-public.yml:116]: This specifies that only individual campaigns can be retrieved, but there's an `all_campaigns` model which indicates the presence of a `list` operation. That should be added in.

File: resources/campaigns/attributes/cmp_is_draft.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+description: Whether or not the campaign is still a draft.
```
Comment (sudoku-lord) [inline resources/campaigns/attributes/cmp_is_draft.yml:1]: See my comment in the campaign model: you don't need to extrapolate these attributes into separate files. You can leave them inline.

File: resources/campaigns/campaigns.yml [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  parameters:
+    - $ref: "../../shared/parameters/limit.yml"
+    - $ref: "../../shared/parameters/include.yml"
+    - $ref: "../../shared/parameters/before_after.yml"
+
+  responses:
+    "200":
+      $ref: "responses/all_campaigns.yml"
+
+  x-codeSamples:
```
Comment (sudoku-lord) [inline resources/campaigns/campaigns.yml:22]: make sure to add in the code samples

File: resources/creatives/models/creative_updatable.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    example: Our updated 4x6 postcard creative
+  from:
+    title: From
+    type: string
+  resource_type:
+    $ref: "../attributes/supported_resource_type.yml"
+  details:
+    title: Details
+    type: object
+  metadata:
```
Comment (sudoku-lord) [inline resources/creatives/models/creative_updatable.yml:16]: https://github.com/lob/lob-openapi/blob/main/shared/models/metadata.yml

Replace metadata with this (and in any other instances where appropriate).

File: resources/creatives/models/creative_writable.yml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    type: string
+    example: Our 4x6 postcard creative
+  resource_type:
+    $ref: "../attributes/supported_resource_type.yml"
+  details:
+    type: object
+  from:
+    title: From
+    type: string
+  metadata:
```
Comment (sudoku-lord) [inline resources/creatives/models/creative_writable.yml:22]: https://github.com/lob/lob-openapi/blob/main/shared/models/metadata.yml see above comment on metadata.

File: resources/uploads/models/uploads.yml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+type: object
+
+properties:
+  foobar:
```
Comment (sudoku-lord) [inline resources/uploads/models/uploads.yml:4]: the request body property is actually called foobar? or is this a placeholder?

**Review by sudoku-lord — CHANGES_REQUESTED** [review]
See comments. I'm particularly curious about why `foobar` is a property in `uploads`. I assume it's a placeholder, but I'd like to know what the intended value is.

File: resources/campaigns/models/campaign.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    properties:
+      id:
+        $ref: "../attributes/cmp_id.yml"
+
+      account_id:
+        description: Account ID that this campaign is associated with.
+        type: string
+
+      is_draft:
+        $ref: "../attributes/cmp_is_draft.yml"
```
[the remark this answers is not in the document]
Comment (jharrell) [reply resources/campaigns/models/campaign.yml]: done

File: resources/campaigns/campaign.yml [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    - Campaigns
+
+  requestBody:
+    required: true
+    content:
+      application/json:
+        schema:
+          $ref: "models/campaign_updatable.yml"
+        example:
+          # TODO: fill out example
```
Comment (BennyKitchell) [inline resources/campaigns/campaign.yml:56]: We have a few `Todo's` here. Do we need examples before merging, or is there a specific JIRA task we can include so this doesn't get lost in the shuffle

File: resources/campaigns/campaigns.yml [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+      application/x-www-form-urlencoded:
+        schema:
+          $ref: "models/campaign_writable.yml"
+        encoding: # be sure to check which encodings apply
+          example-prop:
+            style: deepObject
+            explode: true
+        example:
+          description: encoded demo
+          # complete example using all create payload properties
```
Comment (BennyKitchell) [inline resources/campaigns/campaigns.yml:62]: Same question as above, we should either have examples or add a TODO with a JIRA link so it's not lost.

**Review by BennyKitchell — APPROVED** [review]

File: resources/campaigns/responses/all_campaigns.yml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        - type: object
+          properties:
+            data:
+              type: array
+              description: list of campaigns
+              items:
+                $ref: "../models/campaign.yml"
+
+    example:
+      data:
```
Comment (BennyKitchell) [inline resources/campaigns/responses/all_campaigns.yml:24]: Redocly: `example value must conform to the schema "data" property type must be an array`

File: resources/campaigns/campaigns.yml [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    - $ref: "../../shared/parameters/lang_spec.yml"
+
+  requestBody:
+    required: true
+    content:
+      application/json:
+        schema:
+          $ref: "models/campaign_writable.yml"
+        example:
+          description: "Demo Campaign"
```
Comment (BennyKitchell) [inline resources/campaigns/campaigns.yml:50]: Redocly: `must have required property "name"`

File: resources/campaigns/campaigns.yml (hunk 9 above)
Comment (BennyKitchell) [inline resources/campaigns/campaigns.yml:62]: Redocly: must have required properties "name" and "schedule_type"

File: resources/campaigns/campaigns.yml [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        example:
+          description: encoded demo
+          # complete example using all create payload properties
+          # matching encoding style
+
+      multipart/form-data:
+        schema:
+          $ref: "models/campaign_writable.yml"
+        example:
+          description: demo
```
Comment (BennyKitchell) [inline resources/campaigns/campaigns.yml:69]: Redocly: `must have required properties "name" and "schedule_type"`

File: resources/creatives/creatives.yml [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    - $ref: "../../shared/parameters/lang_spec.yml"
+
+  requestBody:
+    required: true
+    content:
+      application/json:
+        schema:
+          $ref: "models/creative_writable.yml"
+        example:
+          description: "Demo Creative"
```
Comment (BennyKitchell) [inline resources/creatives/creatives.yml:50]: redocly: `Must have required properties "campaign_id" "resource_type" and "details"`

File: resources/creatives/creatives.yml [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+      application/x-www-form-urlencoded:
+        schema:
+          $ref: "models/creative_writable.yml"
+        encoding: # be sure to check which encodings apply
+          example-prop:
+            style: deepObject
+            explode: true
+        example:
+          description: encoded demo
```
Comment (BennyKitchell) [inline resources/creatives/creatives.yml:61]: redocly: `Must have required properties "campaign_id" "resource_type" and "details"`

File: resources/creatives/creatives.yml [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+          # complete example using all create payload properties
+          # matching encoding style
+
+      multipart/form-data:
+        schema:
+          $ref: "models/creative_writable.yml"
+        example:
+          description: demo
```
Comment (BennyKitchell) [inline resources/creatives/creatives.yml:69]: redocly: `Must have required properties "campaign_id" "resource_type" and "details"`

**Pull request merged**